tests/browser/pages/relying-party.js

```diff
@@ -1,39 +1,66 @@
+const axios = require("axios");
+
 module.exports = class PlaywrightDevPage {
   /**
    * @param {import('@playwright/test').Page} page
    */
   constructor(page, clientId) {
     this.page = page;
-    this.clientId = clientId;
 
-    // Starting URL properties
     const websiteHost = process.env.WEBSITE_HOST || "http://localhost:5050";
     this.baseURL = new URL(websiteHost);
+    this.relyingPartyURL = new URL(process.env.RELYING_PARTY_URL);
+    this.env = process.env.ENV || "dev";
 
-    this.oauthPath = `/oauth2/authorize?request=lorem&client_id=${this.clientId}`;
-    this.startingURL = new URL(this.oauthPath, this.baseURL);
-
-    // Relying Party Return URL
-    this.relyingPartyURL = new URL("http://example.net");
+    if (process.env.MOCK_API === "true") {
+      this.oauthPath = this.getOauthPath("lorem", clientId);
+      this.startingURL = new URL(this.oauthPath, this.baseURL);
+    }
   }
 
   async goto() {
+    if (process.env.MOCK_API === "false") {
+      this.startingURL = await this.getStartingURLForStub();
+    }
+
     await this.page.goto(this.startingURL.toString());
   }
 
+  getOauthPath(request, clientId) {
+    return `/oauth2/authorize?request=${request}&client_id=${clientId}`;
+  }
+
+  async getStartingURLForStub() {
+    // needed so that the browser has the credentials set
+    await this.page.goto(this.relyingPartyURL.href);
+
+    const { data } = await axios.get(
+      `${this.relyingPartyURL.href}backend/generateInitialClaimsSet?cri=check-hmrc-${this.env}&rowNumber=197`
+    );
+
+    const {
+      data: { request, client_id },
+    } = await axios.post(
+      `${this.relyingPartyURL.href}backend/createSessionRequest?cri=check-hmrc-${this.env}&rowNumber=197`,
+      data
+    );
+
+    this.oauthPath = this.getOauthPath(request, client_id);
+    return new URL(this.oauthPath, this.baseURL);
+  }
+
   isRelyingPartyServer() {
     const { origin } = new URL(this.page.url());
-
     return origin === this.relyingPartyURL.origin;
   }
 
   hasSuccessQueryParams() {
     const { searchParams } = new URL(this.page.url());
 
     return (
-      searchParams.get("client_id") && // FIXME: Restore checking of client_id
-      searchParams.get("state") === "sT@t3" &&
-      searchParams.get("code").startsWith("auth-code-")
+      !!searchParams.get("client_id") &&
+      !!searchParams.get("state") &&
+      !!searchParams.get("code")
```

Comment:
I like this very much, however; if explicitly returned a nicely named method it would be more apparent for javascript dev(s) not familiar with this idiom, looking at this in future. I think it is fine naming is always hard